### ISSUE TITLE
cli: Checks prevent commands from running in benign situations

### ISSUE THREAD

**Issue by olix0r** [linkerd/linkerd2#1854]
When trying to debug controller issues, it would be helpful to be able to use `linkerd` commands. But, if any of the checks fail, it's impossible to use `linkerd` commands:

```
:; bin/linkerd -l ver-linkerd -n ver-linkerd stat deploy/web                                                                                    
: Error calling Prometheus from the control plane: bad_response: bad response code 500
Validate the install with: linkerd check --linkerd-namespace ver-linkerd
```

`linkerd` should support a `--skip-checks` flag that allows commands to run without running through checks first.

**Comment by klingerf** [comment]
Cool, yeah, it does seem like disabling the CLI when the control plane can't talk to Prometheus is overkill. For my own reference, `linkerd check` runs all of the following checks:

* can initialize the Kubernetes API client
* can query the Kubernetes API
* is running the minimum Kubernetes API version
* control plane namespace exists
* control plane pods are ready
* can initialize the  control plane API client
* can query the control plane API
* control plane can talk to Kubernetes
* control plane can talk to Prometheus
* no invalid service profiles -- _warn only_
* can determine the latest Linkerd version
* cli is up-to-date -- _warn only_
* control-plane is up-to-date -- _warn only_

Whereas running `linkerd dashboard` or some other subcommand only runs this subset:

* can initialize the Kubernetes API client
* can query the Kubernetes API
* control plane namespace exists
* control plane pods are ready
* can initialize the  control plane API client
* can query the control plane API
* control plane can talk to Kubernetes
* control plane can talk to Prometheus

In the example that you pasted, only the last check in that list is failing. In fact, if we can't get past the "can query the control plane API" check, then there's no chance that any other subcommand that talks to the API can succeed. So maybe it would make more sense to just remove those last 2 checks, since they have no bearing on whether or not the CLI can talk to the control plane API.

I also agree that a `--skip-checks` flag could be useful, even if we remove those two problematic checks. I just want to minimize the number of cases where `--skip-checks` is needed.

**Comment by siggy** [comment]
Some of the symptoms of this issue are being addressed in #1899. Agree in principal around some kind of `--skip-checks` flag, but it may be lower priority after #1899 merges.

**Comment by olix0r** [comment]
Furthermore, @cohix reported that he intentionally deleted the `grafana` deploy, but then `dashboard` invocations fail due to checks. We should probably turn a bunch of these failures to warnings.

**Comment by klingerf** [comment]
I agree that deleting the grafana deployment shouldn't prevent users from using the CLI. I think the right fix for this issue is to trim the set of checks that we run prior to executing a CLI command that queries the public API to just the bare minimum set of checks necessary to know that we can at least reach the public API. So modifying the pre-query checks as follows:

* can initialize the Kubernetes API client
* can query the Kubernetes API
* control plane namespace exists
* ~~control plane pods are ready~~
* "linkerd-controller-*" pod is running (not necessarily ready)
* can initialize the  control plane API client
* ~~can query the control plane API~~
* can query the control plane API's `Version` endpoint
* ~~control plane can talk to Kubernetes~~
* ~~control plane can talk to Prometheus~~

If we make those changes, that should substantially reduce the need for a `--skip-checks` flag, but I wouldn't be opposed to adding it anyway if folks feel strong.

**Comment by grampelberg** [comment]
I'm actually a little worried about cutting back. If grafana's not running, clicking on the grafana links in the dashboard feels broken. Maybe we could do something there as well?

**Comment by klingerf** [comment]
@grampelberg That makes sense to me, but `dashboard` feels like a bit of a special case. Should we consider running the reduced set of checks for `tap` / `top`/ `stat` / `routes` / `get`, and the full set of checks for `dashboard`?

**Comment by grampelberg** [comment]
@klingerf that's a great call. Ideally, we'd fit the checks to exactly what those commands need. A close approximation works for me =)
